core-foundation/src/array.rs

```diff
@@ -33,15 +34,47 @@ unsafe impl FromVoid for u32 {
     }
 }
 
-unsafe impl FromVoid for *const c_void {
-    unsafe fn from_void(x: *const c_void) -> *const c_void {
-        x
+unsafe impl<T> FromVoid for *const T {
+    unsafe fn from_void(x: *const c_void) -> Self {
+        x as *const T
     }
 }
 
-unsafe impl FromVoid for CFType {
-    unsafe fn from_void(x: *const c_void) -> CFType {
-        TCFType::wrap_under_get_rule(mem::transmute(x))
+unsafe impl<T: TCFType> FromVoid for T where T::Ref: FromVoid {
+    unsafe fn from_void(x: *const c_void) -> Self {
+        TCFType::wrap_under_get_rule(T::Ref::from_void(x))
+    }
+}
+
+pub struct TCFTypeBorrow<'a, T: 'a>(T, PhantomData<&'a T>);
+
+impl<'a, T> Deref for TCFTypeBorrow<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.0
+    }
+}
+/// A trait describing how to convert from the stored *const c_void to the desired T
+pub unsafe trait FromVoidBorrow {
+    unsafe fn from_void_borrow<'a>(x: *const c_void) -> TCFTypeBorrow<'a, Self> where Self: std::marker::Sized;
+}
+
+unsafe impl FromVoidBorrow for u32 {
+    unsafe fn from_void_borrow<'a>(x: *const c_void) -> TCFTypeBorrow<'a, Self> {
+        TCFTypeBorrow(x as usize as u32, PhantomData)
```

Comment:
I do not understand this casting. Maybe I'm just too tired. There is no dereferencing of the pointer, so would the resulting `u32` not contain the pointer address? Quite likely a truncated version of it as well.

Reply:
Yes, it's intentionally not dereferencing it. I've added a comment to explain it more and dropped the 'as usize' as it doesn't seem to be necessary.